Assets/Mirror/Runtime/ClientObjectManager.cs

```diff
@@ -0,0 +1,685 @@
+            }
+        }
+
+        internal void OnRpcMessage(RpcMessage msg)
+        {
+            if (logger.LogEnabled()) logger.Log("ClientScene.OnRPCMessage hash:" + msg.functionHash + " netId:" + msg.netId);
+
+            if (client.Spawned.TryGetValue(msg.netId, out NetworkIdentity identity))
+            {
+                using (PooledNetworkReader networkReader = NetworkReaderPool.GetReader(msg.payload))
+                    identity.HandleRpc(msg.componentIndex, msg.functionHash, networkReader);
+            }
+        }
+
+        internal void OnSyncEventMessage(SyncEventMessage msg)
+        {
+            if (logger.LogEnabled()) logger.Log("ClientScene.OnSyncEventMessage " + msg.netId);
+
+            if (client.Spawned.TryGetValue(msg.netId, out NetworkIdentity identity))
+            {
+                using (PooledNetworkReader networkReader = NetworkReaderPool.GetReader(msg.payload))
+                    identity.HandleSyncEvent(msg.componentIndex, msg.functionHash, networkReader);
+            }
+            else
+            {
+                logger.LogWarning("Did not find target for SyncEvent message for " + msg.netId);
+            }
+        }
+
+        #endregion
+
+        #region Host
+
+        internal void OnHostClientObjectDestroy(ObjectDestroyMessage msg)
+        {
+            if (logger.LogEnabled()) logger.Log("ClientScene.OnLocalObjectObjDestroy netId:" + msg.netId);
+
+            client.Spawned.Remove(msg.netId);
+        }
+
+        internal void OnHostClientObjectHide(ObjectHideMessage msg)
+        {
+            if (logger.LogEnabled()) logger.Log("ClientScene::OnLocalObjectObjHide netId:" + msg.netId);
+
+            if (client.Spawned.TryGetValue(msg.netId, out NetworkIdentity localObject) && localObject != null)
+            {
+                localObject.OnSetHostVisibility(false);
+            }
+        }
+
+        internal void OnHostClientSpawn(SpawnMessage msg)
+        {
+            if (client.Spawned.TryGetValue(msg.netId, out NetworkIdentity localObject) && localObject != null)
+            {
+                if (msg.isLocalPlayer)
+                    InternalAddPlayer(localObject);
+
+                localObject.HasAuthority = msg.isOwner;
+                localObject.NotifyAuthority();
+                localObject.StartClient();
+                localObject.OnSetHostVisibility(true);
+                CheckForLocalPlayer(localObject);
+            }
+        }
+
+        #endregion
+
+        #region Helpers
+
+        // support additive scene loads:
+        //   * ClientScene.PrepareToSpawnSceneObjects enables them again on the
+        //     client after the server sends ObjectSpawnStartedMessage to client
+        //     in SpawnObserversForConnection. this is only called when the
+        //     client joins, so we need to rebuild scene objects manually again
+        // TODO merge this with FinishLoadScene()?
+        void OnSceneLoaded(Scene scene, LoadSceneMode mode)
```

Comment:
In theory this could fit in with the ClientSceneManager. Might work as an event to call PrepareToSpawnSceneObjects